packages/solana/src/client.ts

```diff
@@ -390,36 +385,47 @@ export class Web3Modal extends Web3ModalScaffold {
   }
 
   private async setProvider(provider: Provider) {
-    const address = await provider.connect()
+    try {
+      this.setLoading(true)
+      const address = await provider.connect()
 
-    const caipChainId = `${SolStoreUtil.state.currentChain?.name}:${SolStoreUtil.state.currentChain?.chainId}`
-    const chain = SolHelpersUtil.getChainFromCaip(
-      this.chains,
-      typeof window === 'object' ? localStorage.getItem(SolConstantsUtil.CAIP_CHAIN_ID) : ''
-    )
-    if (chain) {
-      SolStoreUtil.setCurrentChain(chain)
-    }
-    SolStoreUtil.setIsConnected(true)
-    SolStoreUtil.setCaipChainId(caipChainId)
-    SolStoreUtil.setProvider(provider)
-    this.setAddress(address)
+      // Check if the provider supports the current chain or switch to the first supported chain
+      const connectionChain =
+        provider.chains.find(chain => chain.chainId === SolStoreUtil.state.currentChain?.chainId) ||
+        provider.chains[0]
 
-    window?.localStorage.setItem(SolConstantsUtil.WALLET_ID, provider.name)
+      if (connectionChain) {
+        await this.switchNetwork(
```

Comment:
Why do we do this? Would this be unexpected UX for the users? Is this have some bad cases (ie. like not having a chain provided)?

Reply:
Because the wallet may connect and nothing works because the wallet doesn't support the chain, it is the case of connecting with Trust Wallet and a couple other wallets that only supports mainnet. Then we are just switch to any of the supported chains or stop the connection.